packages/web3/src/contract/contract.ts

```diff
@@ -531,9 +543,32 @@ export class Project {
       }
       return fsPromises.writeFile(artifactPath, compiled.artifact.toString())
     }
-    this.contracts.forEach((contract) => saveToFile(contract))
-    this.scripts.forEach((script) => saveToFile(script))
-    this.saveStructsToFile()
+    for (const [_, contract] of this.contracts) {
+      if (!skipSaveArtifacts || changedSources.find((s) => s.name === contract.sourceInfo.name) !== undefined) {
+        await saveToFile(contract)
+      }
+    }
+    for (const [_, script] of this.scripts) {
+      await saveToFile(script)
+    }
+    await this.saveStructsToFile()
+    await this.saveProjectArtifact(projectRootDir, skipSaveArtifacts, changedSources)
+  }
+
+  private async saveProjectArtifact(projectRootDir: string, skipSaveArtifacts: boolean, changedSources: SourceInfo[]) {
+    if (skipSaveArtifacts) {
+      // we should not update the `codeHashDebug` if the `skipSaveArtifacts` is enabled
```

Comment:
Q: Shall we update `codeHashDebug` when the source code is updated even if `skipSaveArtifacts` is enabled?

Reply:
Yes, this is how it works now.

If the `skipSaveArtifacts` is true, then we will check if the contract code changed by the `changedSource`, if the contract code is changed, then we will use the new `codeHashDebug`.